### PR TITLE
[SHELL32_APITEST] Add CommandLineToArgvW testcase

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Fix command line parsing.
JIRA issue: [CORE-17787](https://jira.reactos.org/browse/CORE-17787)

## Proposed changes

- Add `CommandLineToArgvW` testcase.

## TODO

- [x] Do tests.

### CODE REVIEW

**Comment by katahiromz** [comment]
WinXP:
WinXP
Successful.

Win2k3:
Win2k3
Successful.

Win10:
Win10
Successful.

**Review by ericsomto — APPROVED** [review]

**Review by HBelusca — COMMENTED** [review]
Ah, so the existing **wine**tests for CommandLineToArgvW didn't cover this aspect of the function? (Whatever this aspect is? This is not mentioned anywhere in the PR)

**Comment by katahiromz** [comment]
> didn't cover this aspect of the function?

It didn't cover in `test_commandline2argv` function of `shell32:shlexec` testcase:
https://reactos.org/testman/detail.php?id=62234209&prev=0 .

File: modules/rostests/apitests/shell32/CommandLineToArgvW.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    INT real_argc_minus_1 = real_argc - 1;
+
+    INT m = min(argc_minus_1, real_argc_minus_1);
+    for (INT i = 0; i < m; ++i)
+    {
+        LPCWSTR arg1 = va_arg(va, LPCWSTR);
+        LPCWSTR arg2 = real_argv[i + 1];
+        CStringA str1 = wine_dbgstr_w(arg1);
+        CStringA str2 = wine_dbgstr_w(arg2);
+        ok(lstrcmpW(arg1, arg2) == 0, "Line %d: %s != %s\n", lineno, (LPCSTR)str1, (LPCSTR)str2);
```
Comment (learn-more) [inline modules/rostests/apitests/shell32/CommandLineToArgvW.cpp:28]:
```suggestion
        ok(str1 == str2, "Line %d: %s != %s\n", lineno, (LPCSTR)str1, (LPCSTR)str2);
```

Why did you store it in CString's in the first place?

File: modules/rostests/apitests/shell32/CommandLineToArgvW.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+DoEntry(INT lineno, LPCWSTR cmdline, INT argc_minus_1, ...)
+{
+    va_list va;
+    va_start(va, argc_minus_1);
+
+    INT real_argc;
+    LPWSTR* real_argv = CommandLineToArgvW(cmdline, &real_argc);
+    INT real_argc_minus_1 = real_argc - 1;
+
+    INT m = min(argc_minus_1, real_argc_minus_1);
```
Comment (learn-more) [inline modules/rostests/apitests/shell32/CommandLineToArgvW.cpp:21]: `real_argc_minus_1` is always smaller here?

File: modules/rostests/apitests/shell32/CommandLineToArgvW.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    for (INT i = 0; i < m; ++i)
+    {
+        LPCWSTR arg1 = va_arg(va, LPCWSTR);
+        LPCWSTR arg2 = real_argv[i + 1];
+        CStringA str1 = wine_dbgstr_w(arg1);
+        CStringA str2 = wine_dbgstr_w(arg2);
+        ok(lstrcmpW(arg1, arg2) == 0, "Line %d: %s != %s\n", lineno, (LPCSTR)str1, (LPCSTR)str2);
+    }
+
+    INT diff = abs(argc_minus_1 - real_argc_minus_1);
```
Comment (learn-more) [inline modules/rostests/apitests/shell32/CommandLineToArgvW.cpp:31]: The difference is always 1 here?

File: modules/rostests/apitests/shell32/CommandLineToArgvW.cpp (hunk 1 above)
Comment (katahiromz) [reply modules/rostests/apitests/shell32/CommandLineToArgvW.cpp:28]: `wine_dbgstr_w` looks to have a bug.

File: modules/rostests/apitests/shell32/CommandLineToArgvW.cpp (hunk 3 above)
Comment (katahiromz) [reply modules/rostests/apitests/shell32/CommandLineToArgvW.cpp:31]: Nope. We don't assume it.

File: modules/rostests/apitests/shell32/CommandLineToArgvW.cpp (hunk 2 above)
Comment (katahiromz) [reply modules/rostests/apitests/shell32/CommandLineToArgvW.cpp:21]: Nope. We don't assume it.

Comment (learn-more) [reply modules/rostests/apitests/shell32/CommandLineToArgvW.cpp:21]: Well, line 19 begs to differ I guess.

Comment (katahiromz) [reply modules/rostests/apitests/shell32/CommandLineToArgvW.cpp]: Improved.

File: modules/rostests/apitests/shell32/CommandLineToArgvW.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        ok(str1 == str2, "Line %d: %s != %s\n", lineno, (LPCSTR)str1, (LPCSTR)str2);
+    }
+
+    INT diff = abs(expected_argc - real_argc);
+    while (diff-- > 0)
+    {
+        ok(expected_argc == real_argc,
+           "Line %d: expected_argc %d and real_argc %d are different\n",
+           lineno, expected_argc, real_argc);
+    }
```
Comment (learn-more) [inline modules/rostests/apitests/shell32/CommandLineToArgvW.cpp:34]: Probably easier to just do something like:
`ok(expected_argc == real_argc, "Line %d: expected_argc %d and real_argc %d are different\n", lineno, expected_argc, real_argc);`

**Review by learn-more — APPROVED** [review]
Small simplification, other than that lgtm

File: modules/rostests/apitests/shell32/CommandLineToArgvW.cpp (hunk 4 above)
Comment (katahiromz) [reply modules/rostests/apitests/shell32/CommandLineToArgvW.cpp:34]: To count the number of failures correctly, the original code is better.

**Pull request merged**